### PR TITLE
[app] Add CloudCompute ID serializable within the flow and works state 

### LINKED ISSUE #14803: Add a cloud compute unique identifier serializable within the work state, so it gets tracked.
### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR introduces an internal id used to track CloudCompute within the flow and works state. 

Fixes #14803
Parts of #14800

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

File: src/lightning_app/utilities/packaging/cloud_compute.py [hunk 1]
```diff
@@ -52,9 +55,23 @@ def __post_init__(self):
 
         self.name = self.name.lower()
 
+        # All `default` CloudCompute are identified in the same way.
+        self._internal_id = "default" if self.name == "default" else str(uuid4()).split("-")[0]
```
Comment (ethanwharris) [inline src/lightning_app/utilities/packaging/cloud_compute.py:59]: Safer to use:
```suggestion
        self._internal_id = "default" if self.name == "default" else uuid4().hex
```

**Review by ethanwharris — APPROVED** [review]
LGTM, minor comment

File: src/lightning_app/core/flow.py [hunk 2]
```diff
@@ -177,6 +178,9 @@ def __setattr__(self, name, value):
                 value.component_name = self.name
                 self._state.add(name)
 
+            elif isinstance(value, CloudCompute):
+                self._state.add(name)
```
Comment (awaelchli) [inline src/lightning_app/core/flow.py:182]: @tchaton can you tell me what happens when
```
        self.other_name = CloudCompute(name="gpu")
```
because until now, we had a special property setter/getter. 
Maybe integrate this into your test. A user might do:

```
self.shared_compute = CloudCompute(name="gpu")
```
and in their run method use it to dynamically create work on the same compute.

Comment (tchaton) [reply src/lightning_app/core/flow.py:185]: This works for the flow but not for the works.

File: examples/app_placement/app_v2.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        def get_root():
+            return {"Hello Word!"}
+
+        run(fastapi_service, host=self.host, port=self.port)
+
+
+class Flow(LightningFlow):
+    def __init__(self):
+        super().__init__()
+        self.cloud_compute = CloudCompute(name="cpu-small")
```
Comment (hhsecond) [inline examples/app_placement/app_v2.py:37]: May be we shouldn't show this in the example - adding cloud_compute as an object attribute?

File: src/lightning_app/utilities/packaging/cloud_compute.py [hunk 4]
```diff
@@ -1,5 +1,42 @@
...
+@dataclass
+class _CloudComputeStore:
+    id: str
+    component_names: List[str]
+    frozen: bool = False
+
+    def add_component_name(self, new_component_name: str) -> None:
+        found_index = None
+        for index, component_name in enumerate(self.component_names):
+            if new_component_name.endswith(component_name.replace("root.", "")):
```
Comment (hhsecond) [inline src/lightning_app/utilities/packaging/cloud_compute.py:19]: can you comment here what is this doing? It is be hard for a new reader to figure this out

File: examples/app_placement/app_v2.py (hunk 3 above)
Comment (tchaton) [reply examples/app_placement/app_v2.py:37]: I think this is fine. Why won't you expect CloudCompute to be attached to the flow ?

File: src/lightning_app/utilities/packaging/cloud_compute.py (hunk 4 above)
Comment (tchaton) [reply src/lightning_app/utilities/packaging/cloud_compute.py:19]: We are popping out the previous comment name.

File: src/lightning_app/utilities/packaging/cloud_compute.py [hunk 5]
```diff
@@ -1,5 +1,42 @@
...
+                found_index = index
+
+        if found_index is not None:
+            self.component_names[found_index] = new_component_name
+        else:
+            if len(self.component_names) == 1 and not MULTI_WORKS_INTO_SINGLE_POD and self.id != "default":
+                raise Exception(f"This Cloud Compute is already assigned to {self.component_names[0]}")
```
[the remark this answers is not in the document]
Comment (tchaton) [reply src/lightning_app/utilities/packaging/cloud_compute.py:26]: Done !

File: src/lightning_app/CHANGELOG.md [hunk 6]
```diff
@@ -15,6 +15,11 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Add `--secret` option to CLI to allow binding Secrets to app environment variables when running in the cloud ([#14612](https://github.com/Lightning-AI/lightning/pull/14612))
 
 
+- Add support to upload files to the Drive through an asynchronous `upload_file` endpoint ([#14703](https://github.com/Lightning-AI/lightning/pull/14703))
+
```
Comment (manskx) [inline src/lightning_app/CHANGELOG.md:19]: This is not relevant to this PR?

File: src/lightning_app/core/constants.py [hunk 7]
```diff
@@ -36,7 +36,7 @@
 LIGHTNING_COMPONENT_PUBLIC_REGISTRY = "https://lightning.ai/v1/components"
 LIGHTNING_APPS_PUBLIC_REGISTRY = "https://lightning.ai/v1/apps"
 ENABLE_STATE_WEBSOCKET = bool(int(os.getenv("ENABLE_STATE_WEBSOCKET", "0")))
-
+MULTI_WORKS_INTO_SINGLE_POD = bool(int(os.getenv("MULTI_WORKS_INTO_SINGLE_POD", "0")))
```
Comment (manskx) [inline src/lightning_app/core/constants.py:39]: Let's ensure that the same env_var is used everywhere to enable/disable this feature.

**Review by manskx — APPROVED** [review]

File: src/lightning_app/CHANGELOG.md (hunk 6 above)
Comment (tchaton) [reply src/lightning_app/CHANGELOG.md]: Removed.

File: src/lightning_app/core/constants.py (hunk 7 above)
Comment (tchaton) [reply src/lightning_app/core/constants.py]: Resolved.

File: examples/app_placement/app_v2.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        self.work_a = Work()
```
Comment (lantiga) [inline examples/app_placement/app_v2.py:34]: Since this is an example, I would add comments to make sure readers know what they’re looking at (namely placement, since it’s not very explicit especially with default work)

File: src/lightning_app/core/constants.py [hunk 9]
```diff
@@ -36,8 +36,14 @@
 LIGHTNING_COMPONENT_PUBLIC_REGISTRY = "https://lightning.ai/v1/components"
 LIGHTNING_APPS_PUBLIC_REGISTRY = "https://lightning.ai/v1/apps"
 ENABLE_STATE_WEBSOCKET = bool(int(os.getenv("ENABLE_STATE_WEBSOCKET", "0")))
-
 DEBUG: bool = lightning_cloud.env.DEBUG
+DEFAULT_NUMBER_OF_EXPOSED_PORTS = int(os.getenv("DEFAULT_NUMBER_OF_EXPOSED_PORTS", "50"))
+
+# EXPERIMENTAL: ENV VARIABLE TO ENABLE MULTIPLE WORKS IN A SINGLE POD
+ENABLE_HYBRID = bool(int(os.getenv("ENABLE_HYBRID", "0")))  # Note: This is a disable the feature entirely.
```
Comment (lantiga) [inline src/lightning_app/core/constants.py:43]: I would use a more explicit name, like `ENABLE_HYBRID_BACKEND`

File: src/lightning_app/core/work.py [hunk 10]
```diff
@@ -226,8 +231,17 @@ def cloud_compute(self) -> CloudCompute:
...
-    def cloud_compute(self, cloud_compute) -> None:
+    def cloud_compute(self, cloud_compute: CloudCompute) -> None:
         """Returns the cloud compute used to select the cloud hardware."""
+        # A new ID
+        current_id = self._cloud_compute.id
+        new_id = cloud_compute.id
+        if current_id != new_id:
+            compute_store: _CloudComputeStore = _CLOUD_COMPUTE_STORE[current_id]
+            if compute_store.frozen:
+                raise Exception("The current cloud compute has already been frozen and can't be changed anymore.")
```
Comment (lantiga) [inline src/lightning_app/core/work.py:242]: I’m not sure developers would know what “freezing” is and why we freeze. We can return a more informative exception message that doesn’t rely on our jargon.

**Review by lantiga — APPROVED** [review]
Approved, it would be good if you could address the minor comments there

File: examples/app_placement/app_v2.py (hunk 8 above)
Comment (tchaton) [reply examples/app_placement/app_v2.py:34]: Done !

File: src/lightning_app/core/constants.py (hunk 9 above)
Comment (tchaton) [reply src/lightning_app/core/constants.py:43]: Done !

File: src/lightning_app/core/work.py (hunk 10 above)
Comment (tchaton) [reply src/lightning_app/core/work.py:242]: I removed it as this isn't required anymore with dynamic ports suppport.
